networkx/algorithms/cluster.py

```diff
@@ -247,6 +247,80 @@ def wt(u, v):
         yield (i, dtotal, dbidirectional, float(directed_triangles))
 
 
+@not_implemented_for("directed")
+@nx._dispatchable
+def all_triangles(G, nodes=None):
+    """
+    Yields all unique triangles in an undirected graph using ID assignments and set intersections.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        An undirected graph. Self-loops and multiple edges are ignored.
+
+    nodes : node, iterable of nodes, or None (default=None)
+        If a node or iterable of nodes is given, only triangles involving those nodes are yielded.
+        If None, yields all unique triangles in the graph.
+
+    Yields
+    -------
+    tuple
+        A tuple of three nodes forming a triangle (u, v, w), ordered based on assigned ID,
+        guaranteed to be unique.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(4)
+    >>> list(all_triangles(G))
+    [(0, 1, 2), (0, 1, 3), (0, 2, 3), (1, 2, 3)]
+    """
+    if nodes is None:
+        node_to_id = {node: i for i, node in enumerate(G.nodes())}
+
+        neighbor_sets = {
+            node: {nbr for nbr in G[node] if node_to_id[nbr] > node_to_id[node]}
+            for node in G.nodes()
+        }
+        for u in G.nodes():
+            for v in neighbor_sets[u]:
+                common = neighbor_sets[u] & neighbor_sets[v]
+                for w in common:
+                    yield u, v, w
```

Comment:
This is consuming n^2 extra memory. You don't need to cache the nodes, for you could do:

```suggestion
        for u in G:
            for v in G[u]:
                if node_to_id[v] <= node_to_id[u]:
                    continue
                for w in G[u] & G[v]:
                    if node_to_id[w] <= node_to_id[v]:
                        continue
                    yield (u, v, w)
```

Reply:
Yeah you are right, but for a large graph will they have similar performance?

Reply:
Yes, it could even be faster due to less memory allocations

Reply:
Thank you for the suggestion, I went on and applied it!